### PR TITLE
Make conf json redundant

### PR DESCRIPTION
**Pull request by ronykoz**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/24412

## Description
Building conf.json in build time instead of updating it manually - still allowing to use it.

### CODE REVIEW

**Comment by ronykoz** [comment]
I will add UT later, just want to make sure we are ok with this approach first.

File: Tests/scripts/update_conf_json.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import os
+import yaml
+import json
+from distutils.version import LooseVersion
+
+PACKS_DIR = './Packs'
+CONF_JSON_PATH = './Tests/conf.json'
+INTEGRATIONS_DIR = 'Integrations'
+TEST_PLAYBOOKS_DIR = 'TestPlaybooks'
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:9]: why not sdk constants?

File: Tests/scripts/update_conf_json.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    for pack_name in os.listdir(PACKS_DIR):
+        pack_path = os.path.join(PACKS_DIR, pack_name)
+
+        pack_integrations = []
+        pack_test_playbooks = []
+        for dir_name in os.listdir(pack_path):
+            dir_path = os.path.join(pack_path, dir_name)
+
+            if dir_name not in [INTEGRATIONS_DIR, TEST_PLAYBOOKS_DIR]:
+                continue
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:82]: why use `listdir` if you only refer to those two directories?

File: Tests/scripts/update_conf_json.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+            for file_or_dir in os.listdir(dir_path):
+                if os.path.isdir(os.path.join(dir_path, file_or_dir)) and dir_name == INTEGRATIONS_DIR:
+                    inner_dir_path = os.path.join(dir_path, file_or_dir)
+                    for integration_file in os.listdir(inner_dir_path):
+                        is_yml_file = integration_file.endswith('.yml')
+                        file_path = os.path.join(inner_dir_path, integration_file)
+                        if is_yml_file:
+                            if dir_name == INTEGRATIONS_DIR:
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:91]: redundant due to the condition on line 85

File: Tests/scripts/update_conf_json.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                                pack_integrations.append(get_integration_data(file_path))
+                else:
+                    is_yml_file = file_or_dir.endswith('.yml')
+                    file_path = os.path.join(dir_path, file_or_dir)
+                    if is_yml_file:
+                        if dir_name == INTEGRATIONS_DIR:
+                            pack_integrations.append(get_integration_data(file_path))
+
+                        if dir_name == TEST_PLAYBOOKS_DIR:
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:100]: make sure it's not a test script

File: Tests/scripts/update_conf_json.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    return max_fromversion
+
+
+def calc_conf_json_object(integrations, test_playbooks):
+    conf_objects = [{'playbookID': test_playbook} for test_playbook in test_playbooks]
+    fromverion = get_fromversion(integrations)
+    integrations = [integration_id for integration_id, _ in integrations]
+    if len(integrations) == 1:
+        integrations = integrations[0]
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:38]: list of one should also work btw

File: Tests/scripts/update_conf_json.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    for integration_id, fromversion in integrations:
+        if LooseVersion(fromversion) > LooseVersion(max_fromversion):
+            max_fromversion = fromversion
+
+    return max_fromversion
+
+
+def calc_conf_json_object(integrations, test_playbooks):
+    conf_objects = [{'playbookID': test_playbook} for test_playbook in test_playbooks]
+    fromverion = get_fromversion(integrations)
```
Comment (yaakovi) [inline Tests/scripts/update_conf_json.py:35]: I think we also have `fromversion` for the test playbook, not sure if still relevant (we had an integration we two test playbook that differs in a few tasks that appeared only for a specific version)

File: Tests/scripts/update_conf_json.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import os
```
Comment (glicht) [inline Tests/scripts/update_conf_json.py:1]: add the following at the start:
```
#!/usr/bin/env python3
```
and make the file executable. That way it is possible to call the file directly.

**Comment by glicht** [comment]
Looks good. I like the idea. A few remarks:
* I think we should skip the dir: `Packs/DeprecatedContent/TestPlaybooks` and `Packs/NonSupported`.
* We can optimize the script to avoid parsing integration files if there are no test playbooks or the TestPlaybooks dir doesn't exist.
* Add print for each processed pack so it is easy to see that progress is being made
* Add a print of the total time it takes to generate so we can identify easily when this process gets slow and we need to further optimize (it seems that it takes around a minute currently)
* Copy conf.json to the artifacts dir as we do with id_set.json

File: Tests/scripts/update_conf_json.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        json.dump(conf, conf_file, indent=4)
+
+
+def load_test_data_from_conf_json():
+    test_playbooks = []
+    with open(CONF_JSON_PATH) as data_file:
+        conf = json.load(data_file)
+
+    for conf_test in conf['tests']:
+        test_playbooks.append(conf_test['playbookID'])
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply Tests/scripts/update_conf_json.py:64]: couldn't find it, where should i get it from?

[the remark this answers is not in the document]
Comment (anara123) [reply Tests/scripts/update_conf_json.py:71]: Sorry, my bad. It was in `collect_tests_and_content_packs.py` in content repo.

File: Tests/scripts/update_conf_json.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            new_conf_json_objects.extend(calc_conf_json_object(pack_integrations, pack_test_playbooks))
+
+    add_to_conf_json(new_conf_json_objects)
+
+
+if __name__ == '__main__':
+    start_time = datetime.now()
+    run()
+    total_time = datetime.now() - start_time
+    print(f'Total time {total_time}')
```
Comment (glicht) [inline Tests/scripts/update_conf_json.py:127]: Can we add here also total number of tests collected

Comment (ronykoz) [reply Tests/scripts/update_conf_json.py:128]: @glicht Done

File: .circleci/config.yml [hunk 10]
```diff
@@ -98,6 +98,7 @@ jobs:
                 exit 0
             fi
             demisto-sdk create-id-set -o ./Tests/id_set.json
+            python3 ./Tests/scripts/update_conf_json.py
```
Comment (glicht) [inline .circleci/config.yml:101]: Note that if `demisto-sdk create-id-set...` fails then this line will not be called.

Comment (ronykoz) [reply .circleci/config.yml:101]: If the id_set fails then the upload to marketplace/ configure test filter / configure packs to upload will all not work anyway
@glicht

Comment (glicht) [reply .circleci/config.yml:101]: @ronykoz  Thinking about the flow here.  I think we should do this step before the id_set step. So when id set runs it has all data and is able to include the relevant tests in its output.

**Comment by glicht** [comment]
@ronykoz can you modify an integration, playbook and script to use the new setup so we can see that the whole flow works in the build.

**Comment by ronykoz** [comment]
Added to the artifcats

**Comment by ronykoz** [comment]
@glicht did you look at this?

**Comment by glicht** [comment]
Looking good. Can you format that following json print:
```
Added the following objects to the conf.json:
[{"playbookID": "Test - Windows Defender Advanced Threat Protection", "fromversion": "4.5.0", "integrations": ["Windows Defender Advanced Threat Protection"]}, {"playbookID": "Cymulate Immediate Threats", "fromversion": "4.5.0", "integrations": ["Cymulate"]}, {"playbookID": "stripChars - Test", "fromversion": "4.5.0"}, {"playbookID": "GmailTest", "fromversion": "4.5.0", "integrations": ["Gmail"]}]
Total time 0:01:01.204081
```

**Comment by ronykoz** [comment]
@glicht added the format + reverted the changes to the conf.json
Should be good to go

**Comment by ronykoz** [comment]
@glicht can i merge once the build finish?

**Pull request merged**